src/Runner/Runner.php

```diff
@@ -47,68 +63,216 @@ final class Runner
 
     private CacheManagerInterface $cacheManager;
 
-    private bool $isDryRun;
-
     private LinterInterface $linter;
 
     /**
-     * @var \Traversable<\SplFileInfo>
+     * @var ?iterable<\SplFileInfo>
      */
-    private $finder;
+    private $fileIterator;
+
+    private int $fileCount;
 
     /**
      * @var list<FixerInterface>
      */
     private array $fixers;
 
-    private bool $stopOnViolation;
-
     /**
-     * @param \Traversable<\SplFileInfo> $finder
-     * @param list<FixerInterface>       $fixers
+     * @param null|iterable<\SplFileInfo> $fileIterator
+     * @param list<FixerInterface>        $fixers
      */
     public function __construct(
-        \Traversable $finder,
+        RunnerConfig $runnerConfig,
+        ?iterable $fileIterator,
         array $fixers,
         DifferInterface $differ,
         ?EventDispatcherInterface $eventDispatcher,
         ErrorsManager $errorsManager,
         LinterInterface $linter,
-        bool $isDryRun,
         CacheManagerInterface $cacheManager,
-        ?DirectoryInterface $directory = null,
-        bool $stopOnViolation = false
+        ?DirectoryInterface $directory = null
     ) {
-        $this->finder = $finder;
+        $this->runnerConfig = $runnerConfig;
+        $this->fileIterator = $fileIterator;
         $this->fixers = $fixers;
         $this->differ = $differ;
         $this->eventDispatcher = $eventDispatcher;
         $this->errorsManager = $errorsManager;
         $this->linter = $linter;
-        $this->isDryRun = $isDryRun;
         $this->cacheManager = $cacheManager;
         $this->directory = $directory ?? new Directory('');
-        $this->stopOnViolation = $stopOnViolation;
     }
 
     /**
-     * @return array<string, array{appliedFixers: list<string>, diff: string}>
+     * @param iterable<\SplFileInfo> $fileIterator
+     */
+    public function setFileIterator(iterable $fileIterator): void
```

Comment:
if fileIterator is injected via constructor, I believe it's not a clean design to also be able to override it via setter.

can we make it immutable?

Reply:
It's described [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777/files#diff-451d731f59fc56df00f66e1873fb0800438813b1f87bba0ed062dc801ee81a0cR229). Each worker process is initialised once, the `Runner` instance too. We don't know the paths when process is spawned, the list of paths to analyse is sent from parallelisation operator (main fix/check command) in async React communication, when worker receives the request it just updates `Runner` with paths and runs the analysis.

That's why we need a setter, otherwise we would need to initialise new `Runner` on each request sent from server to the worker. We could clone the `Runner` instance in the setter, assign new paths and return it to achieve immutability, but at the expense of performance, because runner would be cloned several times during the process. I am not sure if it's worth it.

As an alternative I considered making `fixFile()` public and calling it directly from the worker while iterating over paths, but I am not sure it would be cleaner since we would nee to leak `LintingResultInterface` to the worker (along with everything that currently is inside `Runner::getFileIterator()`).